### ISSUE TITLE
epoxy_egl_get_current_gl_context_api never returns EGL_OPENGL_ES_API on conformant EGL 1.4

### ISSUE THREAD

**Issue by chadversary** [anholt/libepoxy#25]
If I understand libepoxy correctly, this bug causes libepoxy to dlsym the wrong library during dispatch resolution. Disclaimer: I found this by statically reading code and I haven't run libepoxy yet.

The EGL 1.4 spec, as of 2013-12-04, says:

```
EGL_OPENG_API and EGL_OPENGL_ES_API are interchangeable for all purposes except eglCreateContext.
```

That includes eglGetCurrentContext too.

Mesa doesn't conform here, though some other EGL implementations do. I filed a Mesa bug long ago and forgot about it: https://bugs.freedesktop.org/show_bug.cgi?id=66552. Thanks for reminding me about it.

**Comment by anholt** [comment]
I keep seeing this bug when I go to look at epoxy stuff, and I still haven't figured out what you're saying we're doing wrong.

**Comment by chadversary** [comment]
The problem is that [epoxy_egl_get_current_gl_context()](https://github.com/anholt/libepoxy/blob/d4620b7dd30e5cd398e10d8d5fb0e25f4cc65817/src/dispatch_common.c#L570) incorrectly returns `EGL_OPENGL_API` instead of `EGL_OPENGL_ES_API` on some implementations. The function's method of determining the current context's API is incorrect.

epoxy_egl_get_current_gl_context() incorrectly assumes that eglGetCurrentContext() will return either a valid **desktop OpenGL** context or NULL if the thread's currently bound EGL API is `EGL_OPENGL_API`. Symmetrically, it also assumes that eglGetCurrentContext() will return either a valid **OpenGL ES** context or NULL if the thread's currently bound EGL API is `EGL_OPENGL_ES_API`. But eglBindAPI() doesn't work that way. eglBindAPI() is badly behaving function whose behavior differs across EGL spec versions and EGL implementations. According to the EGL spec, it's legal for eglGetCurrentContext() to return an **OpenGL ES** context when the currently bound EGL API is `EGL_OPENGL_API`, which causes epoxy_egl_get_current_gl_context() to incorrectly return `EGL_OPENGL_API`.

The only reliable way to determine distinguish between an OpenGL and OpenGL ES context, to the best of my knowledge, is to inspect the context's `GL_VERSION` string.

**Comment by yaronct** [comment]
@chadversary: If all implementations were conformant to the standard, "GL_VERSION" could be the right way. However, in PowerVR OpenGL ES implementation, unfortunately, "GL_VERSION" doesn't return a string starting with "OpenGL ES", so this technique doesn't work.

There's only one more option I can think of (other than let the user set the context type explicitly), which is to use "eglQueryContext" with "EGL_CONTEXT_CLIENT_TYPE" as "attribute". I've written it and pushed to [my fork](https://github.com/yaronct/libepoxy) (branch "egl_type"). Plz test it and tell me whether this works for u.

**Comment by andreislavoiu** [comment]
That workaround for PowerVR needs to be reworked, as currently any spec compliant egl implementation that supports both Desktop OpenGL and OpenGL ES is broken with epoxy. Take for example the following test program:
```
#include <stdio.h>
#include <stdlib.h>
#include <err.h>

#include <epoxy/gl.h>
#include <epoxy/egl.h>

EGLDisplay *get_egl_display_or_skip(void)
{
    Display *dpy = XOpenDisplay(NULL);
    EGLint major, minor;
    EGLDisplay *edpy;
    bool ok;

    if (!dpy)
        errx(77, "couldn't open display\n");

    edpy = eglGetDisplay(dpy);
    if (!edpy)
        errx(1, "Couldn't get EGL display for X11 Display.\n");

    ok = eglInitialize(edpy, &major, &minor);
    if (!ok)
        errx(1, "eglInitialize() failed\n");

    return edpy;
}

int main(int argc, char **argv)
{
    static const EGLint config_attribs[] = {
        EGL_NONE
    };
    static const EGLint context_attribs[] = {
        EGL_CONTEXT_CLIENT_VERSION, 2,
        EGL_NONE
    };
    EGLConfig cfg;
    EGLint count;

    if (!eglBindAPI(EGL_OPENGL_ES_API))
        errx(77, "Couldn't initialize EGL with GLES\n");

    EGLDisplay *edpy = get_egl_display_or_skip();

    if (!eglChooseConfig(edpy, config_attribs, &cfg, 1, &count))
        errx(77, "Couldn't get an EGLConfig\n");

    EGLContext ctx = eglCreateContext(edpy, cfg, EGL_NO_CONTEXT, context_attribs);
    if (!ctx)
        errx(77, "Couldn't create a GL context\n");

    eglMakeCurrent(edpy, NULL, NULL, ctx);
    if (epoxy_is_desktop_gl()) {
        errx(1, "GLES2 context creation made a desktop context: %s\n", glGetString(GL_VERSION));
    }

    if (epoxy_gl_version() < 20) {
        errx(1, "GLES2 context creation made a version %f context\n",
             epoxy_gl_version() / 10.0f);
    }

    printf("result: %p\n", glFenceSync(GL_SYNC_GPU_COMMANDS_COMPLETE, 0));

    return 0;
}

```
When ran with mesa it will print:

> GLES2 context creation made a desktop context: OpenGL ES 3.1 Mesa 12.1.0-devel (git-951bf44)

eglinfo:

> EGL API version: 1.5
> EGL vendor string: Mesa Project
> EGL version string: 1.5 (DRI2)
> EGL client APIs: OpenGL OpenGL_ES 
> EGL extensions string:
>     EGL_CHROMIUM_sync_control EGL_EXT_create_context_robustness
>     EGL_EXT_image_dma_buf_import EGL_KHR_cl_event2 EGL_KHR_create_context
>     EGL_KHR_fence_sync EGL_KHR_get_all_proc_addresses
>     EGL_KHR_gl_colorspace EGL_KHR_gl_renderbuffer_image
>     EGL_KHR_gl_texture_2D_image EGL_KHR_gl_texture_3D_image
>     EGL_KHR_gl_texture_cubemap_image EGL_KHR_image EGL_KHR_image_base
>     EGL_KHR_image_pixmap EGL_KHR_no_config_context EGL_KHR_reusable_sync
>     EGL_KHR_surfaceless_context EGL_KHR_wait_sync
>     EGL_MESA_configless_context EGL_MESA_drm_image
>     EGL_MESA_image_dma_buf_export EGL_NOK_swap_region
>     EGL_NOK_texture_from_pixmap EGL_NV_post_sub_buffer
>     EGL_WL_bind_wayland_display
> EGL client extensions string:
>     EGL_EXT_client_extensions EGL_EXT_platform_base
>     EGL_EXT_platform_wayland EGL_EXT_platform_x11
>     EGL_KHR_client_get_all_proc_addresses EGL_MESA_platform_gbm

es2_info:

> EGL_VERSION: 1.5 (DRI2)
> EGL_VENDOR: Mesa Project
> EGL_EXTENSIONS:
>     EGL_CHROMIUM_sync_control, EGL_EXT_create_context_robustness, 
>     EGL_EXT_image_dma_buf_import, EGL_KHR_cl_event2, EGL_KHR_create_context, 
>     EGL_KHR_fence_sync, EGL_KHR_get_all_proc_addresses, EGL_KHR_gl_colorspace, 
>     EGL_KHR_gl_renderbuffer_image, EGL_KHR_gl_texture_2D_image, 
>     EGL_KHR_gl_texture_3D_image, EGL_KHR_gl_texture_cubemap_image, 
>     EGL_KHR_image, EGL_KHR_image_base, EGL_KHR_image_pixmap, 
>     EGL_KHR_no_config_context, EGL_KHR_reusable_sync, 
>     EGL_KHR_surfaceless_context, EGL_KHR_wait_sync, 
>     EGL_MESA_configless_context, EGL_MESA_drm_image, 
>     EGL_MESA_image_dma_buf_export, EGL_NOK_swap_region, 
>     EGL_NOK_texture_from_pixmap, EGL_NV_post_sub_buffer, 
>     EGL_WL_bind_wayland_display
> EGL_CLIENT_APIS: OpenGL OpenGL_ES 
> GL_VERSION: OpenGL ES 3.1 Mesa 12.1.0-devel (git-951bf44)
> GL_RENDERER: Gallium 0.4 on AMD POLARIS10 (DRM 3.3.0 / 4.8.0-gentoo, LLVM 3.9.0)
> GL_EXTENSIONS:
>     GL_EXT_blend_minmax, GL_EXT_multi_draw_arrays, 
>     GL_EXT_texture_filter_anisotropic, GL_EXT_texture_compression_dxt1, 
>     GL_EXT_texture_format_BGRA8888, GL_OES_compressed_ETC1_RGB8_texture, 
>     GL_OES_depth24, GL_OES_element_index_uint, GL_OES_fbo_render_mipmap, 
>     GL_OES_mapbuffer, GL_OES_rgb8_rgba8, GL_OES_standard_derivatives, 
>     GL_OES_stencil8, GL_OES_texture_3D, GL_OES_texture_float, 
>     GL_OES_texture_float_linear, GL_OES_texture_half_float, 
>     GL_OES_texture_half_float_linear, GL_OES_texture_npot, 
>     GL_EXT_texture_sRGB_decode, GL_OES_EGL_image, GL_OES_depth_texture, 
>     GL_OES_packed_depth_stencil, GL_EXT_texture_type_2_10_10_10_REV, 
>     GL_OES_get_program_binary, GL_APPLE_texture_max_level, 
>     GL_EXT_discard_framebuffer, GL_EXT_read_format_bgra, 
>     GL_NV_fbo_color_attachments, GL_OES_EGL_image_external, GL_OES_EGL_sync, 
>     GL_OES_vertex_array_object, GL_OES_viewport_array, 
>     GL_ANGLE_texture_compression_dxt3, GL_ANGLE_texture_compression_dxt5, 
>     GL_EXT_texture_rg, GL_EXT_unpack_subimage, GL_NV_draw_buffers, 
>     GL_NV_read_buffer, GL_NV_read_depth, GL_NV_read_depth_stencil, 
>     GL_NV_read_stencil, GL_EXT_draw_buffers, GL_EXT_map_buffer_range, 
>     GL_KHR_debug, GL_OES_depth_texture_cube_map, GL_OES_surfaceless_context, 
>     GL_EXT_color_buffer_float, GL_EXT_separate_shader_objects, 
>     GL_EXT_shader_integer_mix, GL_EXT_tessellation_point_size, 
>     GL_EXT_tessellation_shader, GL_EXT_base_instance, GL_EXT_copy_image, 
>     GL_EXT_draw_buffers_indexed, GL_EXT_draw_elements_base_vertex, 
>     GL_EXT_gpu_shader5, GL_EXT_polygon_offset_clamp, 
>     GL_EXT_primitive_bounding_box, GL_EXT_shader_io_blocks, 
>     GL_EXT_texture_border_clamp, GL_EXT_texture_buffer, 
>     GL_EXT_texture_cube_map_array, GL_KHR_context_flush_control, 
>     GL_KHR_robust_buffer_access_behavior, GL_OES_copy_image, 
>     GL_OES_draw_buffers_indexed, GL_OES_draw_elements_base_vertex, 
>     GL_OES_gpu_shader5, GL_OES_primitive_bounding_box, GL_OES_sample_shading, 
>     GL_OES_sample_variables, GL_OES_shader_io_blocks, 
>     GL_OES_shader_multisample_interpolation, GL_OES_tessellation_point_size, 
>     GL_OES_tessellation_shader, GL_OES_texture_border_clamp, 
>     GL_OES_texture_buffer, GL_OES_texture_cube_map_array, 
>     GL_OES_texture_stencil8, GL_OES_texture_storage_multisample_2d_array, 
>     GL_EXT_blend_func_extended, GL_EXT_buffer_storage, 
>     GL_EXT_geometry_point_size, GL_EXT_geometry_shader, 
>     GL_OES_geometry_point_size, GL_OES_geometry_shader, 
>     GL_OES_shader_image_atomic, GL_EXT_clip_cull_distance, 
>     GL_MESA_shader_integer_functions

**Comment by andreislavoiu** [comment]
Suggested patch:

```
--- src/dispatch_common.c.old   2016-10-08 12:41:18.099406705 +0300                                                                                                                                                                                                                                                                                                       
+++ src/dispatch_common.c       2016-10-08 12:42:52.883409136 +0300                                                                                                                                                                                                                                                                                                       
@@ -279,14 +279,8 @@                                                                                                                                                                                                                                                                                                                                                      
      * OpenGL ES, we must also check the context type through EGL (we                                                                                                                                                                                                                                                                                                    
      * can do that as PowerVR is only usable through EGL).                                                                                                                                                                                                                                                                                                               
      */                                                                                                                                                                                                                                                                                                                                                                  
-    if (!epoxy_current_context_is_glx()) {                                                                                                                                                                                                                                                                                                                               
-        switch (epoxy_egl_get_current_gl_context_api()) {                                                                                                                                                                                                                                                                                                                
-        case EGL_OPENGL_API:     return true;                                                                                                                                                                                                                                                                                                                            
-        case EGL_OPENGL_ES_API:  return false;                                                                                                                                                                                                                                                                                                                           
-        case EGL_NONE:                                                                                                                                                                                                                                                                                                                                                   
-        default:  break;                                                                                                                                                                                                                                                                                                                                                 
-        }                                                                                                                                                                                                                                                                                                                                                                
-    }                                                                                                                                                                                                                                                                                                                                                                    
+    if (!epoxy_current_context_is_glx() && epoxy_egl_get_current_gl_context_api() == EGL_OPENGL_ES_API)                                                                                                                                                                                                                                                                  
+        return false;                                                                                                                                                                                                                                                                                                                                                    
 #endif                                                                                                                                                                                                                                                                                                                                                                   
                                                                                                                                                                                                                                                                                                                                                                          
     if (api.begin_count)
```

The idea is if epoxy_egl_get_current_gl_context_api can determine it is a GL ES context trust it, otherwise rely on version string comparison. Should leave the behavior on PowerVR as is now while fixing mesa (tested that it actually does).

**Comment by yaronct** [comment]
@andreislavoiu: Thanx for raising the issue and suggesting a patch! Note that this repo of Epoxy is no longer maintained. I try to maintain a [fork](https://github.com/yaronct/libepoxy) of Epoxy. If u can open a pull request to my fork it'll b highly appreciated!

**Comment by andreislavoiu** [comment]
@yaronct I gave you fork a try and it doesn't need any patching. I believe this [commit](https://github.com/yaronct/libepoxy/commit/8d58c890646fc1f43bcab702bb9ed6bae94daefe) fixes it the proper way.

**Comment by yaronct** [comment]
@andreislavoiu: That's great! I really didn't feel like delving again into that s*** :)

Note that my fork, other than being maintained, also contains a number of bug fixes and improvements, notably it can be built as a static or shared library using cmake (while retaining autotools support), including running tests, and doesn't depend on `eglplatform.h` or `khrplatform.h`. If u do encounter any further problem with my fork or wish to improve it in any way plz let me know!

**Comment by chadversary** [comment]
@nwnk submitted a fix as https://github.com/anholt/libepoxy/pull/90

**Comment by yaronct** [comment]
@chadversary, @nwnk: I can't understand what drives u to open a pull request to an unmaintained repository, but it's your choice..

**Comment by nwnk** [comment]
@yaronct Partly force of habit, partly because I actually know anholt and am hoping to shame him into either merging the various branches and doing another release or delegating ownership.

**Comment by yaronct** [comment]
@nwnk: Oh, either would b great, I hope he's shamed enough :laughing:

**Comment by ebassi** [comment]
Merged #90 in master, so I think we can close this.
